doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,183 @@
+Application Persistent Data Storage Permissions
+===============================================
+
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft <br/>
+**Author:** Brad Campbell<br/>
+**Draft-Created:** 2024/06/06<br/>
+**Draft-Modified:** 2024/06/06<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** devel@lists.tockos.org<br/>
+
+
+Abstract
+-------------------------------
+
+Tock supports storing persistent state for applications, and all persistent
+state in Tock is identified based on the application that stored it. Tock
+supports permissions for persistent state, allowing for the kernel to restrict
+which applications can store state and which applications can read stored state.
+This TRD describes the permissions architecture for persistent state in Tock.
+This document is in full compliance with [TRD1][TRD1].
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Stored State Identifiers
+-------------------------------
+
+All shared persistent storage implementations must store a 32 bit identifier
+with each stored object to mark the application that created the stored object.
+
+When applications write data, their ShortId must be used as the identifier. When
+the kernel writes data, the identifier must be 0.
```

Comment:
I wonder if `0` is the right choice here.

While I suspect metadata / storage are versioned more explicitly in the actual storage implementation, there's an implicit statement here about the instance of the Tock kernel which stored this thing. If we have signed kernel images, or other versions of 'secure/proven' kernels [i.e., an ID that says, "this is OpenTitan's official kernel"], but then we later have to revise / update / etc the mechanism that protects / proves kernel ids for whatever reason, how should the new kernel know if it should trust the "kernel data" in storage?

Or even the simpler question—how should a 'signed/trusted/verified' kernel know whether the stored data was written by an 'official' kernel image?

Hmm... I'm talking myself out of this thinking through the actual security properties more; maybe that's something that needs to be pushed to a lower layer.

---

Moving past the half complete thoughts above, I'll just leave this as a question—is there value in having a "kernel ID"? Even something as simple as an "ID" of `0x54 0x4b 0x02 0x01` (`TK21` for Tock Kernel version 2.1)?

Reply:
Yeah `0x00` is a reserved storage permission as well. It would be tricky to try and reserve others: https://book.tockos.org/doc/tock_binary_format#7-storage-permissions